File: app/src/main/java/ru/vegax/xavier/a3test/repositories/UsersRepository.java
package ru.vegax.xavier.a3test.repositories;

import android.arch.lifecycle.MutableLiveData;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;
import ru.vegax.xavier.a3test.models.User;
import ru.vegax.xavier.a3test.models.UserAlbum;
import ru.vegax.xavier.a3test.models.UserAlbumWrapper;
import ru.vegax.xavier.a3test.models.UserPhoto;
import ru.vegax.xavier.a3test.models.UserPhotoWrapper;
import ru.vegax.xavier.a3test.models.UserWrapper;
import ru.vegax.xavier.a3test.retrofit2.TypicodeApi;

//Singleton pattern
public class UsersRepository {
    private static UsersRepository instance;
    private final MutableLiveData<UserWrapper> userData = new MutableLiveData<>();
    private final MutableLiveData<UserAlbumWrapper> userAlbumData = new MutableLiveData<>();
    private final MutableLiveData<UserPhotoWrapper> userPhotoData = new MutableLiveData<>();
    private static TypicodeApi typicodeApi;

    public static UsersRepository getInstance(){
        if(instance == null){
             instance = new UsersRepository();
        }
        Retrofit retrofit = new Retrofit.Builder().baseUrl("https://jsonplaceholder.typicode.com/")
                .addConverterFactory(GsonConverterFactory.create()).build();



        typicodeApi = retrofit.create(TypicodeApi.class);

        return instance;
    }
    public MutableLiveData<UserWrapper> getUsers(){

        Call<List<User>> userCall = typicodeApi.getUsers();
        userCall.enqueue(new Callback<List<User>>() {
            @Override
            public void onResponse(@NotNull Call<List<User>> call, @NotNull Response<List<User>> response) {
                if (!response.isSuccessful()) {
                    //send error to UI
                    userData.setValue(new UserWrapper(new Throwable("could not load code: " + response.code())));                           return;
                }
                List<User> myUserAlbums = response.body();
                userData.setValue(new UserWrapper(myUserAlbums));
            }

            @Override
            public void onFailure(@NotNull Call<List<User>> call, @NotNull Throwable t) {
                //send error to UI
                userData.setValue(new UserWrapper(new Throwable("could not load code: " + "no connection  " + t.getMessage() )));
            }
        });
        return userData;
    }
    public MutableLiveData<UserAlbumWrapper> getUserAlbums(){

        Call<List<UserAlbum>> userCall = typicodeApi.getUserAlbums();
        userCall.enqueue(new Callback<List<UserAlbum>>() {
            @Override
            public void onResponse(@NotNull Call<List<UserAlbum>> call, @NotNull Response<List<UserAlbum>> response) {
                if (!response.isSuccessful()) {
                    //send error to UI
                    userAlbumData.setValue(new UserAlbumWrapper(new Throwable("could not load code: " + response.code())));                           return;
                }
                List<UserAlbum> myUserAlbums = response.body();
                userAlbumData.setValue(new UserAlbumWrapper(myUserAlbums));
            }

            @Override
            public void onFailure(@NotNull Call<List<UserAlbum>> call, @NotNull Throwable t) {
                //send error to UI
                userAlbumData.setValue(new UserAlbumWrapper(new Throwable( "could not load code: " + "no connection  " + t.getMessage())) );
            }
        });
        return userAlbumData;
    }

    public MutableLiveData<UserPhotoWrapper> getUserPhoto(){

        Call<List<UserPhoto>> userCall = typicodeApi.getUserPhotos();
        userCall.enqueue(new Callback<List<UserPhoto>>() {
            @Override
            public void onResponse(@NotNull Call<List<UserPhoto>> call, @NotNull Response<List<UserPhoto>> response) {
                if (!response.isSuccessful()) {
                    //send error to UI
                    userPhotoData.setValue(new UserPhotoWrapper(new Throwable("could not load code: " + response.code())));
                    return;
                }
                List<UserPhoto> myUserPhotos = response.body();
                userPhotoData.setValue(new UserPhotoWrapper(myUserPhotos));
            }

            @Override
            public void onFailure(@NotNull Call<List<UserPhoto>> call, @NotNull Throwable t) {
               //send error to UI
                userPhotoData.setValue(new UserPhotoWrapper(new Throwable("could not load code: " + "no connection  " + t.getMessage() )));
            }
        });
        return userPhotoData;
    }
}

File: app/src/main/java/ru/vegax/xavier/a3test/repositories/UsersRepository.kt
package ru.vegax.xavier.a3test.repositories

import android.arch.lifecycle.MutableLiveData

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import ru.vegax.xavier.a3test.models.User
import ru.vegax.xavier.a3test.models.UserAlbum
import ru.vegax.xavier.a3test.models.UserAlbumWrapper
import ru.vegax.xavier.a3test.models.UserPhoto
import ru.vegax.xavier.a3test.models.UserPhotoWrapper
import ru.vegax.xavier.a3test.models.UserWrapper
import ru.vegax.xavier.a3test.retrofit2.TypicodeApi

//Singleton pattern
class UsersRepository {
    private val userData = MutableLiveData<UserWrapper>()
    private val userAlbumData = MutableLiveData<UserAlbumWrapper>()
    private val userPhotoData = MutableLiveData<UserPhotoWrapper>()
    //send error to UI
    //send error to UI
    val users: MutableLiveData<UserWrapper>
        get() {

            val userCall = typicodeApi?.users
            userCall?.enqueue(object : Callback<List<User>> {
                override fun onResponse(call: Call<List<User>>, response: Response<List<User>>) {
                    if (!response.isSuccessful) {
                        userData.value = UserWrapper(Throwable("could not load code: " + response.code()))
                        return
                    }
                    val myUserAlbums = response.body()
                    userData.value = UserWrapper(myUserAlbums)
                }

                override fun onFailure(call: Call<List<User>>, t: Throwable) {
                    userData.value = UserWrapper(Throwable("could not load code: " + "no connection  " + t.message))
                }
            })
            return userData
        }
    //send error to UI
    //send error to UI
    val userAlbums: MutableLiveData<UserAlbumWrapper>
        get() {

            val userCall = typicodeApi!!.userAlbums
            userCall.enqueue(object : Callback<List<UserAlbum>> {
                override fun onResponse(call: Call<List<UserAlbum>>, response: Response<List<UserAlbum>>) {
                    if (!response.isSuccessful) {
                        userAlbumData.value = UserAlbumWrapper(Throwable("could not load code: " + response.code()))
                        return
                    }
                    val myUserAlbums = response.body()
                    userAlbumData.value = UserAlbumWrapper(myUserAlbums)
                }

                override fun onFailure(call: Call<List<UserAlbum>>, t: Throwable) {
                    userAlbumData.value = UserAlbumWrapper(Throwable("could not load code: " + "no connection  " + t.message))
                }
            })
            return userAlbumData
        }

    //send error to UI
    //send error to UI
    val userPhoto: MutableLiveData<UserPhotoWrapper>
        get() {

            val userCall = typicodeApi!!.userPhotos
            userCall.enqueue(object : Callback<List<UserPhoto>> {
                override fun onResponse(call: Call<List<UserPhoto>>, response: Response<List<UserPhoto>>) {
                    if (!response.isSuccessful) {
                        userPhotoData.value = UserPhotoWrapper(Throwable("could not load code: " + response.code()))
                        return
                    }
                    val myUserPhotos = response.body()
                    userPhotoData.value = UserPhotoWrapper(myUserPhotos)
                }

                override fun onFailure(call: Call<List<UserPhoto>>, t: Throwable) {
                    userPhotoData.value = UserPhotoWrapper(Throwable("could not load code: " + "no connection  " + t.message))
                }
            })
            return userPhotoData
        }

    companion object {
        private var instance: UsersRepository? = null
        private var typicodeApi: TypicodeApi? = null

        fun getInstance(): UsersRepository {
            if (instance == null) {
                instance = UsersRepository()
            }
            val retrofit = Retrofit.Builder().baseUrl("https://jsonplaceholder.typicode.com/")
                    .addConverterFactory(GsonConverterFactory.create()).build()



            typicodeApi = retrofit.create(TypicodeApi::class.java)

            return instance as UsersRepository
        }
    }
}
